addons/stock/models/stock_inventory.py

```diff
@@ -74,7 +74,7 @@ def copy_data(self, default=None):
 
     def unlink(self):
         for inventory in self:
-            if inventory.state not in ('draft', 'cancel'):
+            if inventory.state not in ('draft', 'cancel') and not self.env.context.get('_force_unlink', False):
```

Comment:
by RPC it is possible to unlink inventory, add a condition to check if it is an super user.

@odony, I think that `_force_unlink`, it is used in many place (for uninstall). It could be secure that `_force_unlink` can only use by an admin user.

Solution : add this condition here:
odoo/model.py
```python
def unlink(self)
   if self.env.context.get('_force_unlink', False) and not self.user.is_admin():
       raise UserError(_('Not allowed to perform this action'))
   ...
```

Reply:
All RPC-callable methods that directly or indirectly set `_force_unlink` / `MODULE_UNINSTALL_FLAG` seem to be decorated with `assert_log_admin_access` which does what you suggest, do you have a specific use-case where this is problematic?

Reply:
In this case @Elkasitu it is better to use `MODULE_UNINSTALL_FLAG` than `_force_unlink`.

And I have find nowhere "MODULE_UNINSTALL_FLAG seem to be decorated with assert_log_admin_access"

Reply:
I guess @fmdl's point is that anyone can send a crafted RPC request to directly `unlink()` with `_force_unlink` in the context of the call, without being an admin.

It's true, although our business rules are generally meant to prevent accidental deletions by regular users, not malicious manually crafted RPC requests. Such an aggressive deletion would still be logged for audit purposes, and the user who is doing that is purposely violating the terms of his employment contract.
I'm not sure we want to do anything about this in the code. There are many legit ways to make a mess in a database if you're looking for trouble anyway.

Still, using the `MODULE_UNINSTALL_FLAG` constant seems more readable. And perhaps we do want to look for some generic solution to this uninstallation problem.

Reply:
> And perhaps we do want to look for some generic solution to this uninstallation problem.
I agree, I've had this in the back of my mind for a while now, maybe it's time to write some specs.

> Still, using the `MODULE_UNINSTALL_FLAG` constant seems more readable.
Will do, however I will also investigate if it can be turned into a registry property or something like that for ease of access, importing `odoo.addons.base.ir.ir_model` just for this flag is annoying, in some cases can create a dependency cycle and is not even cross-compatible between 11 and 12

Reply:
idea : add the flag in env

if self.env.flag_unistall:
    ...

Reply:
@odony it is not a big issue today, because `MODULE_UNINSTALL_FLAG` is not use a lot. Maybe in the furtur this flag will be use more and security issue can appear.
`MODULE_UNINSTALL_FLAG` is a bypass, it could be normal to reduce the access only for admin people.